TESTS/mbed_platform/FileHandle/main.cpp

```diff
@@ -0,0 +1,532 @@
+    TEST_ASSERT_EQUAL_INT(str1_size, write_ret);
+
+#ifndef __ICCARM__ // prevents IAR infinite loop
+    // write 3; expected written 2
+    TestFile<FS>::resetFunctionCallHistory();
+    write_ret = std::fwrite(str2, 1, str2_size, file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnWrite));
+    TEST_ASSERT_TRUE(std::ferror(file) > 0);
+    std::clearerr(file); // for ARMCC
+#ifndef __ARMCC_VERSION
+    // ARMCC returns 0 here instead of number of elements successfully written
+    TEST_ASSERT_EQUAL_INT(str2_size - 1, write_ret);
+#endif
+
+    // write 3; expected written 0
+    TestFile<FS>::resetFunctionCallHistory();
+    write_ret = std::fwrite(str1, 1, str1_size, file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnWrite));
+    TEST_ASSERT_TRUE(std::ferror(file) > 0);
+    std::clearerr(file); // for ARMCC
+    TEST_ASSERT_EQUAL_INT(0, write_ret);
+#endif
+
+    std::fseek(file, 0, SEEK_SET);
+
+    // read 3; expected read 3
+    TestFile<FS>::resetFunctionCallHistory();
+    read_ret = std::fread(read_buf, 1, str1_size, file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnRead));
+    TEST_ASSERT_EQUAL_INT(str1_size, read_ret);
+    TEST_ASSERT_EQUAL_INT(0, strncmp(str1, read_buf, str1_size));
+
+#ifndef __ICCARM__
+    // read 3; expected read 2
+    TestFile<FS>::resetFunctionCallHistory();
+    read_ret = std::fread(read_buf, 1, str2_size, file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnRead));
+    TEST_ASSERT_TRUE(std::feof(file) > 0);
+    std::clearerr(file); // for ARMCC
+    TEST_ASSERT_EQUAL_INT(str2_size - 1, read_ret);
+    TEST_ASSERT_EQUAL_INT(0, strncmp(str2, read_buf, str2_size - 1));
+
+    // read 3; expected read 0
+    TestFile<FS>::resetFunctionCallHistory();
+    read_ret = std::fread(read_buf, 1, str2_size, file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnRead));
+    TEST_ASSERT_TRUE(std::feof(file) > 0);
+    std::clearerr(file); // for ARMCC
+    TEST_ASSERT_EQUAL_INT(0, read_ret);
+#endif
+
+    std::fclose(file);
+}
+
+/** Test fputc and fgetc
+ *
+ *  Given already opened file
+ *
+ *  When write some data to file
+ *  Then underneath retargeting layer write function is called
+ *       fputc return written element
+ *       on failure, returns EOF and stream error is sets
+ *
+ *  When read previously written data from file
+ *  Then underneath retargeting layer read function is called
+ *       fgetc return read element
+ *       read data match previously written
+ *       on failure, returns EOF and stream error/eof is sets respectively
+ *
+ */
+void test_fputc_fgetc()
+{
+    std::FILE *file;
+    const uint32_t FS = 3;
+    TestFile<FS> fh;
+    char char_buf[3] = {'a', 'b', 'c' };
+    int ret;
+
+    file = fdopen(&fh, "w+");
+    TEST_ASSERT_NOT_NULL(file);
+#ifdef __ICCARM__
+    std::setbuf(file, NULL); // fix IAR every byte sync
```

Comment:
Well, this then wouldn't need an ifdef. mbed_set_unbuffered_stream just calls `std::setbuf(file, NULL)` if not ICCARM.

Reply:
You mean in #5571 ? On master there is still `std::setvbuf(_file,buf,_IONBF,NULL);`

Reply:
If I'm not confused, this expands to

    #ifdef __ICC_ARM__
        setbuf(file, NULL);
    #else
        // mbed_set_unbuffered_stream(file):
        #if defined (__ICCARM__)
          char buf[2];
          setvbuf(file,buf,_IONBF,NULL);
        #else
          setbuf(file, NULL);
      #endif
    #endif

Which, more consisely is

    setbuf(file, NULL);

Reply:
Now get it:) My idea was to preserve `mbed_set_unbuffered_stream(file);` to show that this should be fixed by #5571 or in the future

So truncate it to single line `setbuf(file, NULL);` ?

Reply:
I think so - the `set_unbuffered_stream` (and `mbed_gets` and `mbed_getc`) are just some sort of IAR workaround for the normal functions `setbuf` `gets` and `getc` at some point not working properly - a few bits of code like `Stream` knew to call them instead. If we are currently thinking `setbuf` is working, just use it and drop the workaround.

Reply:
fixed